Clarify and link cloud/security goals (#1651)
sourcegraph/about: handbook/engineering/security/index.md

1. Create alerts and dashboards to automate the process of investigating events of interest (e.g., detect and alert on a spike of failed login attempts to a single account, or across Sourcegraph Cloud as a whole).

### Confidence in our security model

**Problem:** We need to have a high degree of confidence in our overall security before we ask paying customers to trust us with their private code.

**Milestones:**

1. [Visibility into Sourcegraph Cloud's attack surface](#visibility-into-sourcegraph-clouds-attack-surface).
   - I suspect we don't need to block on everything in that goal. Maybe we just need milestones 1-3? Let's discuss.
1. We connect our [test security repository](https://github.com/sourcegraph/security-test/blob/main/README.md) to Sourcegraph Cloud and only [members who can access that repository on GitHub](https://github.com/sourcegraph/security-test/settings/access) can access that repository on Sourcegraph Cloud (i.e., [Sourcegraph organization owners](https://github.com/orgs/sourcegraph/people?query=role%3Aowner) and [@sourcegraph/security](https://github.com/orgs/sourcegraph/teams/security) members).

> 1. We advertise a bounty of \$1000 for each unique vulnerability that allows an unauthorized person to gain access to our [test security repository](https://github.com/sourcegraph/security-test/blob/main/README.md) on Sourcegraph Cloud.

Author: ```suggestion
1. We advertise a bounty for each unique vulnerability that allows an unauthorized person to gain access to our [test security repository](https://github.com/sourcegraph/security-test/blob/main/README.md) on Sourcegraph Cloud.
```

> 1. We run a time-bound capture the flag event where there is a single large bounty (e.g., \$10,000) for the first person who reports a vulnerability that allows them to gain access to our [test security repository](https://github.com/sourcegraph/security-test/blob/main/README.md) on Sourcegraph Cloud.

Author: ```suggestion
1. We run a time-bound capture the flag event where there are larger bounties for being able to gain access to our [test security repository](https://github.com/sourcegraph/security-test/blob/main/README.md) on Sourcegraph Cloud.
```

## Goals

Our current work is documented in our [tracking issue](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aissue+label%3Atracking+label%3Ateam%2Fsecurity+is%3Aopen).

### Visibility into Sourcegraph Cloud's attack surface

**Problem:** The Cloud team is working toward [private code on Sourcegraph Cloud](../cloud/index.md#private-code-on-sourcegraph-cloud). Before we can have [confidence in our security model](#confidence-in-our-security-model), we first need to have visibility into current risks, issues, and threats, so we can proactively plan what work needs to be done (i.e., we don't want to be relying on or waiting for security researchers or customer to report vulnerabilities to us).

**Milestones:**

> 1. All Docker images are continuously scanned for known security vulnerabilities and the security team is alerted any time one is found.

Reviewer: All Docker images are continuously scanned for known security vulnerabilities and the security team is alerted as vulnerabilities are found.

Author: (if you use the suggestion format then I can directly apply your suggestions)

```suggestion
1. All Docker images are continuously scanned for known security vulnerabilities and the security team is alerted as vulnerabilities are found.
```

> 1. All compute nodes are continuously scanned for known security vulnerabilities and the security team is alerted any time one is found.

Reviewer: All compute nodes are continuously scanned for known security vulnerabilities and the security team is alerted as vulnerabilities are found.

Author: ```suggestion
1. All compute nodes are continuously scanned for known security vulnerabilities and the security team is alerted as vulnerabilities are found.
```

>    - GCP doesn't handle this?

Author: ```suggestion
```

> 1. All high risk (I mean this informally, is there a better term of art to use here?) vulnerabilities found in scans are resolved.

Reviewer: Vulnerabilities with a high severity, or a CVSS score of at least 4.0 are resolved.

Author: ```suggestion
1. Vulnerabilities with a high severity, or a CVSS score of at least 4.0 are resolved.
```

1. Implement centralized security logging.

>    - What is this? Is this just taking all our existing logs and shipping them to one place? Are there new kinds of logs we will want to add to our application? If so, what/where? We are already using LogDNA, is that a solution to this?

Reviewer 2: We should not be using LogDNA — that was an experiment that @keegancsmith started. I think we should keep our logs, which may contain sensitive data, inside our core infrastructure (i.e GCP), so we should use https://console.cloud.google.com/logs/viewer?project=sourcegraph-dev

Reviewer: We wanted store security logs (to start) in Hosted ElasticSearch to get started. The other side of the same discussion is storing (also Elastic backed) in our own infrastructure, but creating a new GCP project separate from production and dev.

Reviewer 2: I think we need to align on one centralized logging solution for engineering at large (including security use-cases), so this needs to be discussed. Will there be an RFC?

Reviewer 3: yeah please link me to this rfc. As a sidenote we have had pretty bad experiences with elastic search for logging in the past. Currently (and for quite a while) I imagine the amount of log data we generate can just be searched with grep and we can avoid solutions that require maintaining an index.

Reviewer: Can I ask what the issues were that were experienced in the past? Was it speed, of search, managing Elastic itself, of something else entirely?

Reviewer 3: All 3. It was slow, it was tricky to update configuration for it to parse our structured logs and it would constantly die. All in all I don't understand using it for logs at our scale (or even much larger scales). If you are generating a few GB of logs a day, that is totally reasonable to just grep. We also won't be using extra features kibana provides like alerting, we should be sticking to prometheus for that.

Maybe someone with more experience it it could make it work well for us, but to be honest this seems like a problem that shouldn't require someone with specific experience :P

Reviewer: So we used it to build a large data pipeline for security logs, as did the folks over at [siemmonster](https://siemonster.com/community-edition/).  We had it up to about 100GB of logs a day across, with closing indices, and structuring. But, we were using logstash to prep the data in the first place.

I fully agree - Elastic experience makes a difference in making it behave rationally.  However, the security side does care about the alerting. Our goal/use case is to both have the centralized logs, and work down the path of correlation.

Author: Implementation details should be discussed in a relevant RFC when we get to that milestone. For now, it is enough to say that we want to have a place where we can search over all of our logs.

1. Centralize workload and cloud (i.e. GCE) logs.

>    - Needs to be explained more.

Reviewer: Centralize workload and cloud (i.e. GCE) logs; this provides visibility from a systems and services perspective and can be used for investigating and correlating security events of interest.

> 1. Normalize log format across services so it is easier to correlate and search, and this is a precursor to creating automated alerts from the logs.

Reviewer: Normalize log format across services so that it is easier to correlate and search; this is a precursor to creating automated alerts from the logs.

Author: ```suggestion
1. Normalize log format across services so that it is easier to correlate and search; this is a precursor to creating automated alerts from the logs.
```

1. Add support for audit and access logging (e.g., visibility into both intentional and unintentional logins). This is both useful from a security visibility point of view, as a requirement for various auditing frameworks (though not currently a target).
1. Create alerts and dashboards to automate the process of investigating events of interest (e.g., detect and alert on a spike of failed login attempts to a single account, or across Sourcegraph Cloud as a whole).

### Confidence in our security model

**Problem:** We need to have a high degree of confidence in our overall security before we ask paying customers to trust us with their private code.

**Milestones:**

1. [Visibility into Sourcegraph Cloud's attack surface](#visibility-into-sourcegraph-clouds-attack-surface).

>    - I suspect we don't need to block on everything in that goal. Maybe we just need milestones 1-3? Let's discuss.

Reviewer 4: I would suspect 1-4 would be sufficient to unblock, but 5 & 6 seem closely coupled with 4, so it might be reasonable to do all of them together. I know @chayim wants to focus on logging relatively soon.

Author: I think the key point is that we find and resolve all the easy/obvious things before we start asking people to attack us and rewarding large bounties.

It seems like "adding centralized logging" won't actually yield new insights unless we are under active attack, so that is why I think it isn't a pre-requisite to starting to award bounties (the next goal). Do you agree or disagree?

Author: I updated the PR and added the reasons there. If you disagree, please comment on the new diff.
